go/cmd/vttablet/vttablet.go

```diff
@@ -138,5 +142,22 @@ func main() {
 		// to update our state, so closing it in OnClose()
 		ts.Close()
 	})
+
+	if *tabletBackupAndExit {
+		go func() {
```

Comment:
To make sense as a feature of vttablet, I would expect backup-and-exit to integrate cleanly with vttablet's own lifecycle code. For example, it should know how to avoid registering in topology, avoid running queryservice, synchronously wait for restore and replication to catch up, take the backup, and exit cleanly with exit code 0.

I'm not saying those aspects are required to make a useful backup-and-exit tool; I'm saying without those, it doesn't make sense for the tool to be built into vttablet. In this current form, we aren't gaining anything that couldn't be done more flexibly from outside vttablet by a bash script: start vttablet and mysqlctld, poll health and replication delay endpoint (either gRPC or just fetching /debug/status), trigger backup, terminate vttablet and mysqlctld.

Do you see any reason the bash script won't work? If not, can we just start with bash and come back to the upstream patch when we think we can do something that improves on the bash script (e.g. in the ways mentioned at the top of this comment)?

Reply:
The bash script can work, but to get it to work the bash script is going to have to manage multiple processes, handle polling for status, and handle triggering and waiting for the backup to complete. This change is 10 lines that cleanly adds this workflow. 

Eventually it will be good to have this process not use as much of the vttablet code ,but any solution that anyone would use today is going to be more complex then this, require more moving parts and more configuration ( since it would need to integrate with vtctld ), and is going to do the same thing. I think it makes sense to incorporate this directly into vttablet for the time being and pull it out once we have a better solution

Reply:
I still feel this implementation (sitting in a loop outside the real tablet lifecycle) is too ad hoc to be fitting for inclusion upstream. Without integrating natively in the lifecycle, there are likely to be things that could go wrong if someone uses it in a scenario that doesn't meet implicit assumptions that are only valid for your particular environment.

Given that we know there's a better way and you just want to do this for expediency, I suggest it should be a patch that you carry internally until someone has time to work on a general solution that's suitable for many users. @sougou do you have an opinion on this?